js/reaction.js

```diff
@@ -170,11 +171,42 @@ function listen(node) {
   $('.integertext', node).on('blur', event => checkInteger(event.target));
 }
 
+/**
+ * Clicks the 'save' button if ready for a save.
+ */
+function clickSave() {
+  // Only save if there are unsaved changes still to be saved -- hence save
+  // button visible -- and if ready for a save (not in the process of saving
+  // already).
+  const saveButton = $('#save');
+  if (saveButton.css('visibility') == 'visible' &&
+      saveButton.text() == 'save') {
+    saveButton.trigger('click');
+  }
+}
+
 /**
  * Shows the 'save' button.
  */
 function dirty() {
   $('#save').css('visibility', 'visible');
+  // Handle timers for autosave.
+  // Short timer restarts on modification, in order to wait until a sufficiently
+  // long duration of no change.
+  if (session.timers['short']) {
+    clearTimeout(session.timers['short']);
+  }
+  session.timers['short'] =
+      setTimeout(clickSave, 1000 * 5);  // Save after five seconds
+  // Long timer saves a while after each modification, but not too frequently.
+  // This is done by only starting the timer if it's not already active.
+  if (!session.timers['long']) {
```

Comment:
I think the documentation of these timers is confusing; AFAICT the `long` timer doesn't care whether there are recent modifications; it just saves on a regular interval.

Reply:
Made changes to documentation; explaining this well has been pretty tough. If it's still not great, will keep thinking about it.

Reply:
I think you're making it more complicated than it needs to be.

There are two timers. One saves the form at regular intervals. The other resets on each modification and saves after a short period of inactivity. Having both timers means that the form still gets saved regularly while it is being populated, without a long delay after the last modification.

Reply:
Sorry that I'm misunderstanding something...

I think as it's coded right now, the long timer doesn't do quite that -- it does depend on recent changes. (Eg if it saves, and you don't make any changes after, the timer will be completely off and save attempts won't happen. The timer won't start again til the next change.)

Should I change this to simply save periodically (like what you described)?

Reply:
I guess I don't understand the purpose of the second timer if it is reset by modifications just like the first.

Reply:
If the second timer's counting down as a change is made, it's not reset -- it continues to run. I think that this way, during a long period of constant modifications we can still get a save; and, during periods of no modifications, we won't have saves.

Perhaps a simple periodic timer may be good enough? (especially with already checking for no extraneous saves)

Reply:
> If the second timer's counting down as a change is made, it's not reset -- it continues to run. I think that this way, during a long period of constant modifications we can still get a save; and, during periods of no modifications, we won't have saves.

That was my understanding, and it justifies the need for the second timer to have saves after briefer periods of inactivity.